### PR TITLE
Add `ext_iterator_interface_compat`, a portable alternative to `iterator_interface`.

### LINKED ISSUE #10: Address deducing *this dependency
This code, unlike the implementation used in beman.optional26, depends on deducing *this, a C++23 feature. This dependency makes it impossible to evaluate the library in projects that are using C++20 and earlier toolchains.

We could conditionally compile an implementation that doesn't depend on deducing *this, but this changes the interface. With deducing this, note that the CRTP isn't used,

```c++
template <class IteratorConcept,
          class ValueType,
          class Reference      = ValueType&,
          class Pointer        = ValueType*,
          class DifferenceType = ptrdiff_t>
class iterator_interface;
```

With the Boost, version, the CRTP is used. Note the `Derived` template argument.

```c++
template <typename Derived,
          typename IteratorConcept,
          typename ValueType,
          typename Reference      = ValueType&,
          typename Pointer        = ValueType*,
          typename DifferenceType = std::ptrdiff_t
          >
struct iterator_interface;
```

Here are some possible ways to address this:

1. **Do nothing**. Here we turn away many potential users and opportunities for feedback.
2. **Make a second structure available**. Here we'd provide a `iterator_interface_crtp` which doesn't use deducing *this.
3. **Provide different interfaces based on a build setting**. In this case we'd have a configuration variable that would select whether the CRTP or deducing this interface is available.

### PR DESCRIPTION
**Pull request by camio**
The `iterator_interface` class template requires deducing *this support which came in C++23 and is not yet widely available. This PR introduces an alternative class template `ext_iterator_interface_compat` which requires an additional template argument (for RAII support), but otherwise has the same functionality of `iterator_interface` without the deducing *this dependency.

* The new class was prefixed with `ext_` to make it clear that this is an extension to the `iterator_interface` paper. It may be interesting to adopt this as a convention for Beman libraries in general.
* Added a build option, `BEMAN_ITERATOR_INTERFACE26_USE_DEDUCING_THIS`, that
indicates whether deducing *this is available. This element of configuration is placed in a generated `config.hpp`. This was chosen, as opposed to using a define passed as a compiler flag, to avoid ODR violations with larger projects with incoherent flag usage.
* If deducing *this is available, `ext_iterator_interface_compat` is implemented with `iterator_interface`. If it is not available, instantiation of `iterator_interface` results in a helpful error message and `ext_interator_interface_compat` is implemented using an in-source copy of `Boost.Stl_Interfaces` with the following modifications:
   * It was relicenced to "Apache-2.0 WITH LLVM-exception" per permission from Zach Laine
   * It uses `beman::iterator_interface26::iterator_interface_access` for access control for consistency.
* Testing support was modified to follow the pattern described in https://github.com/beman-project/exemplar/issues/60#issuecomment-2452092354
  
The intent is to add documentation in a separate PR once this gets moved in.

Fixes #10.

### CODE REVIEW

File: include/beman/iterator_interface26/config.hpp.in [hunk 1]
```diff
@@ -0,0 +1,6 @@
+#ifndef BEMAN_ITERATOR_INTERFACE26_CONFIG_HPP
```
Comment (neatudarius) [inline include/beman/iterator_interface26/config.hpp.in:1]: Would you like to remove 26 from codebase or from this PR (diff)?

File: include/beman/iterator_interface26/detail/stl_interfaces/config.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
+// include/beman/optional26/detail/stl_interfaces/config.hpp -*-C++-*-
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// Copyright (C) 2020 T. Zachary Laine
```
Comment (neatudarius) [inline include/beman/iterator_interface26/detail/stl_interfaces/config.hpp:4]: I suppose this is code from boost.  Just a reminder to check and keep only want we need.

File: include/beman/iterator_interface26/config.hpp.in (hunk 1 above)
Comment (camio) [reply include/beman/iterator_interface26/config.hpp.in:1]: I intend to leave 26 in PR (and remove it separately). What I *do* want to remove in this PR is all references to *optional*.

File: include/beman/iterator_interface26/detail/stl_interfaces/config.hpp (hunk 2 above)
Comment (camio) [reply include/beman/iterator_interface26/detail/stl_interfaces/config.hpp:4]: This was taken from optional26 so I think it's already been trimmed down.

**Review by steve-downey — APPROVED** [review]
I don't think this was part of this change, but
https://github.com/beman-project/iterator_interface/pull/13/files#diff-7d7f79f7e81b3a5127eb407b43c39c90befb00593ef4d6721ee28edf4e87f9f9R179-R184
is now generating a dangling temporary warning?

**Comment by camio** [comment]
> is now generating a dangling temporary warning?

Added #15 to address this. I couldn't figure out the cause with an initial look.

**Review by neatudarius — COMMENTED** [review]
LGTM+

**Pull request merged**